### PR TITLE
Soql integration with apex grammar

### PR DESCRIPTION
**Pull request by kevinTveizaga**
Partial integration with Apex grammar. soql of type [select...] its supported now.

### CODE REVIEW

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexKeyword.java [hunk 1]
```diff
@@ -123,7 +123,39 @@
...
+    //Where Notations
+    WHERE("WHERE"),
+    NOT_SOQL("NOT"),
+    AND_SOQL("AND"),
+    OR_SOQL("OR"),
+    IN("IN"),
+    INCLUDES("includes"),
+    EXCLUDES("excludes"),
+    LIKE("LIKE"),
+    //Limit Notations
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexKeyword.java:144]: Did you forget add the LIMIT notation?

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/SOQLExpressions.java [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+                grammarBuilder.optional(NOT_SOQL),
+                grammarBuilder.firstOf(
+                        IN,
+                        INCLUDES,
+                        EXCLUDES),
+                LPAREN,
+                grammarBuilder.firstOf(SOQL_NAME, QUERY_EXPRESSION),
+                grammarBuilder.zeroOrMore(COMMA, SOQL_NAME),
+                RPAREN
+
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/SOQLExpressions.java]: remove blank lines

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/SOQLExpressions.java [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        filteringExpressions(grammarBuilder);
+        operatorToComparisson(grammarBuilder);
+        conditionalSOQLExpression(grammarBuilder);
+        andSOQLExpression(grammarBuilder);
+        orSOQLExpression(grammarBuilder);
+        limitSentence(grammarBuilder);
+        orderBySentence(grammarBuilder);
+        groupBySentence(grammarBuilder);
+        groupByTypesSentence(grammarBuilder);
+
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/SOQLExpressions.java]: remove blank lines

File: apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarQueryExpressionTest.java [hunk 4]
```diff
@@ -0,0 +1,461 @@
+/*
```
Comment (JimmyRojas) [inline apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarQueryExpressionTest.java:1]: invalid header class

File: apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarQueryExpressionTest.java [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+    }
+
+    @Test
+    public void positiveRules_ConditionalOperatorsWhereQueryExpression() {
+        assertThat(parser)
+                .matches("SELECT field1,field2 "
+                        + "FROM table1 AS T1 WHERE customer_id = 1 LIMIT 100")
+                .matches("SELECT field1,field2 "
+                        + "FROM table1 AS T1 WHERE Name LIKE 'A%' LIMIT 100")
+                .matches("SELECT field1,field2, "
```
Comment (JimmyRojas) [inline apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarQueryExpressionTest.java]: it looks the same as Line 137

File: apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarQueryExpressionTest.java [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+    }
+
+    @Test
+    public void positiveRules_ComplexWhereOrQueryExpression() {
+        assertThat(parser)
+                .matches("SELECT field1,field2 FROM table1 AS T1 "
+                        + "WHERE customer_id = 1 OR id = 1 LIMIT 100")
+                .matches("SELECT field1,field2 FROM table1 AS T1 "
+                        + "WHERE Name LIKE 'A%' OR C = 5 LIMIT 100")
+                .matches("SELECT field1,field2, "
```
Comment (JimmyRojas) [inline apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarQueryExpressionTest.java]: it looks the same as Line 181

File: apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarSOQLExpressionTest.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (JimmyRojas) [inline apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarSOQLExpressionTest.java]: invalid header class

File: apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarSOQLExpressionTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.Test;
+import org.junit.Before;
+
+import org.fundacionjala.enforce.sonarqube.apex.parser.ApexRuleTest;
+
+import static org.fundacionjala.enforce.sonarqube.apex.api.grammar.ApexGrammarRuleKey.SOQL_EXPRESSION;
+import static org.sonar.sslr.tests.Assertions.assertThat;
+
+public class ApexGrammarSOQLExpressionTest extends ApexRuleTest {
+/**
```
Comment (JimmyRojas) [inline apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarSOQLExpressionTest.java]: you should not add commented code

File: apex-squid/src/test/resources/complex/ComplexClass.cls [hunk 9]
```diff
@@ -27,9 +27,9 @@ public with sharing class ComplexClass
 			{
 				MetadataService.MetadataPort service = MetadataServiceExamples.createService();		
 				List<MetadataService.ListMetadataQuery> queries = new List<MetadataService.ListMetadataQuery>();		
-				MetadataService.ListMetadataQuery query = new MetadataService.ListMetadataQuery();
-				query.type_x = node;
-				queries.add(query);		
+				MetadataService.ListMetadataQuery query1 = new MetadataService.ListMetadataQuery();
```
Comment (JimmyRojas) [inline apex-squid/src/test/resources/complex/ComplexClass.cls]: improve the variable name

File: apex-squid/src/test/resources/complex/ComplexClassViatra.cls [hunk 10]
```diff
@@ -1,6 +1,6 @@
 public with sharing class ADDR_Validator_Batch implements Database.Batchable<sObject>, Database.AllowsCallouts {
 
-    private String query;
+    private String query1;
```
Comment (JimmyRojas) [inline apex-squid/src/test/resources/complex/ComplexClassViatra.cls]: improve the variable name

**Comment by JimmyRojas** [comment]
see a few comments there

**Comment by JimmyRojas** [comment]
I think we need to add the grammar for:

Dynamic SOQL - https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_dynamic_soql.htm

e.g.
String myTestString = 'TestName';
List<sObject> sobjList = Database.query('SELECT Id FROM MyCustomObject__c WHERE Name = :myTestString');

e.g.
Account[] accts = [SELECT Name,Phone FROM Account];

see more samples here, https://developer.salesforce.com/trailhead/en/apex_database/apex_database_soql

**Comment by VarS311** [comment]
@JimmyRojas Here's the update on the SOQL part, I merged it with the latest changes and added the limit check

File: apex-checks/src/test/resources/checks/soqlLimit.cls [hunk 11]
```diff
@@ -0,0 +1,12 @@
+public class SomeClass {
+	
+	public static void aMethod() {
+            string incorrect = [SELECT dato FROM table1];
```
Comment (JimmyRojas) [inline apex-checks/src/test/resources/checks/soqlLimit.cls]:
```
string incorrect = [SELECT dato FROM table1];
List<table1> incorrect = [SELECT dato FROM table1];
```

just to be consistent with SFDC

File: apex-checks/src/test/resources/checks/soqlLimit.cls [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+        }
+
+        public static void aMethod() {
+            boolean correct = [SELECT dato FROM table1 LIMIT 50];
```
Comment (JimmyRojas) [inline apex-checks/src/test/resources/checks/soqlLimit.cls]:
```
boolean correct = [SELECT dato FROM table1 LIMIT 50];
List<table1> correct = [SELECT dato FROM table1 LIMIT 50];
```

just to be consistent with SFDC

**Comment by JimmyRojas** [comment]
just a couple of comments

File: apex-checks/src/test/resources/checks/soqlLimit.cls (hunk 11 above)
Comment (VarS311) [inline apex-checks/src/test/resources/checks/soqlLimit.cls]: Ok, will fix it right away

**Comment by JimmyRojas** [comment]
Looks good to me, thank you.

**Pull request merged**